### PR TITLE
Fix dark and white theme switch in Chrome

### PR DESCRIPTION
**Pull request by florianduros**
## Checklist

* [X] Tests written for new code (and old code if feasible)
* [X] Linter and other CI checks pass
* [X] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

Notes: Fixes (vector-im/element-web/issues/22609) where the white theme is not applied when `white -> dark -> white` sequence is done.

Chrome fires the `load` event of a style sheet only the first time when the sheet is enabled. If the sheet is disabled and enabled, the sheet will not fire a new `load` event.  

To reproduce, go to element-web with the white theme selected by default:
 - Switch to dark => the dark theme is applied
 - Switch dark to white => the dark theme is still active
 

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fixes (vector-im/element-web/issues/22609) where the white theme is not applied when `white -> dark -> white` sequence is done. ([\#9320](https://github.com/matrix-org/matrix-react-sdk/pull/9320)). Contributed by @florianduros.

### CODE REVIEW

**Comment by t3chguy** [comment]
Merging the branch won't make the CI pass, it needs to get the minimum test coverage in order

image

File: src/theme.ts [hunk 1]
```diff
@@ -237,8 +237,8 @@ export async function setTheme(theme?: string): Promise<void> {
 
     // look for the stylesheet elements.
     // styleElements is a map from style name to HTMLLinkElement.
-    const styleElements = Object.create(null);
-    const themes = Array.from(document.querySelectorAll('[data-mx-theme]'));
+    const styleElements: Record<string, HTMLLinkElement> = Object.create(null);
```
Comment (weeman1337) [inline src/theme.ts:240]: Can you please switch this to a `Map` here?
Background is that we shouuld avoid `Record` types for key-based lookups + a `Map` may be faster.

**Comment by florianduros** [comment]
> Merging the branch won't make the CI pass, it needs to get the minimum test coverage in order
> 
> image

I have failed ci job  about the login (`login.spec.ts`). I'm a bit confused, I think I didn't touch the login code.

In which job, I can see that error about code coverage ? 
I'll add some unit tests but I'm afraid about theirs pertinence in this use case.

Thanks

**Review by weeman1337 — CHANGES_REQUESTED** [review]
Just one comment. Can you add tests, update the branch and then re-request review?

**Comment by t3chguy** [comment]
@florianduros https://sonarcloud.io/summary/new_code?id=matrix-react-sdk&pullRequest=9320 

image

File: test/theme-test.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                    disabled: true,
+                    href: 'urlDark',
+                    onload: () => void 0,
+                },
+            ];
+            lightTheme = styles[0];
+            darkTheme = styles[1];
+
+            jest.spyOn(document.body, 'style', 'get').mockReturnValue([] as any);
+            jest.spyOn(document, 'querySelectorAll').mockReturnValue(styles as any);
```
Comment (weeman1337) [inline test/theme-test.ts:51]: I would use `mockImplementation` here and check for the expected query.

**Review by weeman1337 — APPROVED** [review]
Thanks for adding the tests. Tested on the Netlify deployment with Chrome, but I don't have access to an environment as described in the issue. To be 100 % someone with this env should test it.

File: test/theme-test.ts (hunk 2 above)
Comment (florianduros) [reply test/theme-test.ts:51]: The expected query can be checked as well with `spyOn`, which is used to create mock function and track the calls

Comment (florianduros) [reply test/theme-test.ts]: I pushed a check for the expected query :)

**Comment by florianduros** [comment]
> 

Which environnement ?

**Comment by weeman1337** [comment]
> Which environnment ?

Chrome on Windows 11

**Comment by turt2live** [comment]
Right, so I've done some testing on Windows 10 and 11 with not-great results. It seems to get stuck in dark theme for some reason, and on Win 11 Edge specifically it doesn't seem to come out of dark theme:

(regedit because these are not activated versions of Windows in the slightest)

https://user-images.githubusercontent.com/1190097/193131817-1d7f60c0-2c51-43ea-974f-92de898f4425.mp4


https://user-images.githubusercontent.com/1190097/193131822-80b0ca3e-a8f6-4703-adf2-1821dec445b0.mp4

**Review by weeman1337 — CHANGES_REQUESTED** [review]
Thanks to Travis for testing!

@florianduros can you have another look at it and re-request review once you have finished?

**Comment by florianduros** [comment]
I tested in w11 VM and I don't have the issue of @turt2live 

https://user-images.githubusercontent.com/2621378/194333793-d93a5a33-2bd1-431b-aa18-7060741e4491.mov

Chrome on the left and Edge on the right

**Comment by weeman1337** [comment]
I've tested this on the Windows 10 dev VM using Chrome. Works fine there. Also tested the current development and can reproduce the issue there. @florianduros can you update your branch?

**Review by weeman1337 — APPROVED** [review]

**Pull request merged**